Shopify/ruby-lsp
# Add autocomplete for classes, modules and constants

### Motivation

Closes #61 

Add completion for classes, modules and constants.

### Implementation

I'll explain the changes per file
1. `Executor`: when a completion request is triggered, we receive the position for the character immediately after the letter the user just typed. So I started checking if the previous character (the one just inserted) is an uppercase character and then we go down the path of completion for classes, modules and constants. Otherwise, we do the previous path completion. Also started passing the `nesting` to the completion request and adapted our capabilities
2. `Completion`: started handling `const` and `const_path_ref` to provide completion for both namespaced and regular constants. These events use the `prefix_search` added in #955 and then build the completions. I also extracted the logic to build markdown content from hover into `Common`, so that we can display the exact same documentation on both hover and completion (the documentation shows up on the side when selecting a completion)

Other than that, I just renamed the request to be more generic, using `Completion` instead of `PathCompletion`.

### Automated Tests

Added tests for both const and const path refs.

### Manual Tests

1. Open a codebase that does not use Sorbet (e.g.: rails/rails)
2. Configure the Ruby LSP to use this branch `"rubyLsp.branch": "vs/add_autocomplete"`
3. Type some uppercase letter somewhere
4. Verify you get completion options
5. Verify you can see the documentation properly when switching options (notice that in VS Code the documentation might be hidden. You need to click the little right arrow on top of a completion item to show docs)
